src/odemis/acq/stitching/_registrar.py

```diff
@@ -14,18 +14,534 @@
+        t2 = max(0, min(self.size[0], - shift[1]))
+        r2 = max(0, min(self.size[1], self.size[1] - shift[0]))
+        b2 = max(0, min(self.size[0], self.size[0] - shift[1]))
+
+        return (l1, t1, r1, b1), (l2, t2, r2, b2)
+
+    def _estimateMatch(self, imageA, imageB, shift=(0, 0)):
+        """
+        Returns an estimation of the similarity between the given images
+        when the second is shifted by the shift value. It is used to assess 
+        the quality of a shift measurement by giving the shifted image.
+        return (0 <= float<=1): the bigger, the more similar are the images
+        """
+        # If case the shift is extreme, force the match to 0, to indicate
+        # something is wrong
+        if math.hypot(*shift) >= self.tsize + self.osize * EXTREME_SHIFT:
+            return 0
+
+        (l1, t1, r1, b1), (l2, t2, r2, b2) = self._estimateROI(shift)
+
+        imageA_sh = numpy.array(imageA)[t1:b1, l1:r1]
+        imageB_sh = numpy.array(imageB)[t2:b2, l2:r2]
+
+        avg = (numpy.sum(imageA_sh) / imageA_sh.size,
+               numpy.sum(imageB_sh) / imageB_sh.size)
+
+        dist = (numpy.subtract(imageA_sh, avg[0]),
+                numpy.subtract(imageB_sh, avg[1]))
+
+        covar = numpy.sum(dist[0] * dist[1]) / imageA_sh.size
+
+        var = (numpy.sum(numpy.power(dist[0], 2)) / imageA_sh.size,
+               numpy.sum(numpy.power(dist[1], 2)) / imageB_sh.size)
+
+        stDev = (math.sqrt(var[0]), math.sqrt(var[1]))
+
+        if stDev[0] == 0 or stDev[1] == 0:
+            return 0
+
+        return covar / (stDev[0] * stDev[1])
+
+    def _get_shift(self, prev_tile, tile, shift=(0, 0)):
+        """
+        Measures the shift on the overlap between the two tiles when the second is shifted \
+        by the shift value.
+        shift (2 ints): shift between prev_tile and tile
+        return (2 ints): guessed shift
+        """
+        (l1, t1, r1, b1), (l2, t2, r2, b2) = self._estimateROI(shift)
+
+        a = prev_tile[t1:b1, l1:r1]
+        b = tile[t2:b2, l2:r2]
+
+        [x, y] = MeasureShift(b, a)
+        return x, y
+
+    def _mean_shift(self, row, col, coords_array):
+        """
+        Returns the mean shift computed along the given row and column
+        """
+        val = [coords_array[row][i] for i in range(1, col) if coords_array[row][i] is not None] + \
+              [coords_array[i][col]
+                  for i in range(1, row) if coords_array[i][col] is not None]
+        if not val:
+            raise ValueError("Not enough data to compute mean horizontal shift")
+        mean_x = int(sum(v[0] for v in val) / len(val))
+        mean_y = int(sum(v[1] for v in val) / len(val))
+        return mean_x, mean_y
+
+
+    def _pos_to_left_right(self, row, col, xdir):
+        """
+        Returns the position of the new tile after calculating the shift with the
+        neighbor on the left or right
+        xdir: 1 - left, -1 - right
```

Comment:
`xdir` values are only `LEFT` or `RIGHT`, don't mention the values of these constants here.